File: python/delete_middle_node.py
from linked_list import LinkedList

# time complexity: O(1), space complexity: O(1)
def delete_middle_node(node):
    if node == None or node.next == None:
        return False
    next_node = node.next
    node.data = next_node.data
    node.next = next_node.next
    return True


list = LinkedList(1)
list.add(2)
list.add(3)
list.add(4)
list.add(5)
list.add(6)
list.add(7)
list.add(8)
list.add(9)

node = list.head.next.next.next
delete_middle_node(node)
list.print_data()

File: java/DeleteMiddleNode.java
// Implement an algorith to delete a node in the middle of a singly linked list
// given only access to that node

class DeleteMiddleNode {
    public static void main(String args[]) {
        MyLinkedList<Integer> list = new MyLinkedList<Integer>(1);
        list.add(2);
        list.add(3);
        list.add(4);
        list.add(5);
        list.add(6);
        list.add(7);
        list.add(8);
        list.add(9);

        MyNode node = list.head.next.next.next;
        deleteMiddleNode(node);
        list.printData();
    }

    // time complexity: O(1), space complexity: O(1)
    private static boolean deleteMiddleNode(MyNode node) {
        if (node == null || node.next == null) {
            return false;
        }
        node.data = node.next.data;
        node.next = node.next.next;
        return true;
    }
}

